qkeras/quantizers.py

```diff
@@ -439,9 +443,517 @@ def trainable_variables(self):
+        (https://arxiv.org/pdf/1502.02551.pdf).
+      scale_axis (int, None): Which axis to calculate scale from. If None, we
+        perform per-channel scaling based off of the image data format. Note 
+        that each entry of a rank-1 tensor is considered its own channel by
+        default. See `_get_scaling_axis` for more details. Defaults to None.
+      qnoise_factor (float): A scalar from 0 to 1 that represents the level of
+        quantization noise to add. This controls the amount of the
+        quantization noise to add to the outputs by changing the weighted
+        sum of (1 - qnoise_factor) * unquantized_x + qnoise_factor *
+        quantized_x. Defaults to 1.0, which means that the result is fully
+        quantized.
+      use_variables (bool): If true, we use tf.Variables to store certain 
+        parameters. See the BaseQuantizer implementation for more details.
+        Defaults to False. If set to True, be sure to use the special attribute
+        update methods detailed in the BaseQuantizer.
+      var_name (str or None): A variable name shared between the tf.Variables
+        created in on initialization, if use_variables is true. If None, the 
+        variable names are generated automatically based on the parameter names 
+        along with a uid. Defaults to None.
+
+    Returns:
+      function: Function that computes linear quantization.
+
+    Raises:
+      ValueError:
+        - If `bits` is not positive, or is too small to represent `integer`.
+        - If `integer` is negative.
+        - If `alpha` is a string but not one of ("auto", "auto_po2").
+
+    """
+
+  # string options for alpha parameter
+  ALPHA_OPTIONS = ("auto", "auto_po2")
+
+  def __init__(
+      self,
+      bits=8,
+      integer=0,
+      symmetric=1,
+      keep_negative=True,
+      alpha=None,
+      use_stochastic_rounding=False,
+      scale_axis=None,
+      qnoise_factor=1.0,
+      var_name=None,
+      use_variables=False,
+  ):
+    super(quantized_linear, self).__init__()
+
+    # Set _initialized parameter to False to prevent the setters from
+    # performing preliminary calculations
+    self._initialized = False
+    self.var_name = var_name
+    self.bits = bits
+    self.integer = integer
+    self.symmetric = symmetric
+    self.keep_negative = keep_negative
+    self.qnoise_factor = qnoise_factor
+    self.use_stochastic_rounding = use_stochastic_rounding
+    self.alpha = alpha
+    self.scale_axis = scale_axis
+    self.use_variables = use_variables
+
+    # Perform preliminary calculations based on attributes above
+    self._initialized = True
+    self._set_default_quantization_scale()
+
+
+  @property
+  def bits(self):
+    return self._bits
+
+  @bits.setter
+  def bits(self, bits):
+    self._bits = bits
+    err_msg = f"Bit count {bits} must be positive"
+    if bits <= 0:
+      raise ValueError(err_msg)
+    if self._initialized:
```

Comment:
The purpose of this is to update the quantization scale (and in effect, the scale) immediately when any attributes are changed. If we make these attributes non-modifiable, then this is not necessary

Reply:
It's useful to have the quantization_scale updated automatically since it is used in the min, max and range functions. The scale is also used in qtools, so that's of course important too

Reply:
The risk of updating quantization_scale automatically is that training will potentially fluctuate greatly without ever converging. 

That's why setting it as weight which get updated through gradient descent is probably a better solution.

Reply:
The other thing, is _set_default_quantization_scale a proper name? If you are calling this function every time to update quantization_scale, then it's not returning a default value?

Reply:
See the new commits. Most of the parameters are now non-modifiable, and the _set_default_quantization_scale method is now a default_quantization_scale property